### DISCUSSION TITLE
Alpine Expression Error when trying to use dynamic import

### DISCUSSION THREAD

**Discussion by Briareos17** [alpinejs/alpine#3635] [7. Bugs (archived)] [answered]
I'm trying dynamic import of my components, it looks like:

Component code:

```
"use strict";

export default (options) => {
    options = {
        ...options,
        ...{
            hidden: false
        }
    }

    return {
        hidden: options.hidden,
        init() {
            this.observeFooter()
        },
        scrollToTop() {
            window.scrollTo({
                top: 0,
                behavior: 'smooth'
            });
        },
        observeFooter() {
            const observer = new IntersectionObserver(entries => {
                const [entry] = entries;

                this.toggleVisibility(entry.isIntersecting);
            }, {
                threshold: 0,
                rootMargin: '0px 0px 200px 0px'
            });

            const footer = document.getElementById('footer');

            observer.observe(footer);
        },
        toggleVisibility(state) {
            this.hidden = !state;
        }
    }
}
```

Import code:

```
Alpine.data('ScrollToTopButton', () => {
    return import('./components/ScrollToTopButton');
});
```

But, on page load I've got this error:

```
module.esm.js:421 Alpine Expression Error: hidden is not defined
Expression: "!hidden"
```

it because I'm using `x-show="!hidden"` in component but variable `hidden` not initialized yes in case of dynamic import, is there any way to fix this?

**Comment by ekwoka** [comment]
dynamic import returns a promise.

Alpine has no ability to wait for that promise.

**Reply by Briareos17** [reply]
Is there any plans to make this possible, I'm not so familiar with JS, but isn't it just to check promise and call evalueation of directives after x-data evaluated?

**Reply by ekwoka** [reply]
It would take a significant rewrite of the internals. Since the evaluation of the tree in all synchronous.

You just need to have all your components available before running Alpine.start().

Best to just do it in module scripts, probably with just static import.

**Comment by Tim-Wils** [comment]
There’s this third party library https://async-alpine.dev/docs/ maybe that’s what you’re looking for?

**Reply by ekwoka** [reply]
Yes, that can. What it basically does is run through the dom before Alpine does and use x-ignore or just removing attributes to prevent alpine from initializing those components until a future time.

**Reply by Briareos17** [reply]
This is more looks like a crutch than good practice

**Reply by Tim-Wils** [reply]
Probably wouldn't use it myself either, because it looks to bloated for my likings. 
I think I'll use a new directive that's something like this (pseudocode, didn't test it):
```
Alpine.directive('async-data', async (el, { value, modifiers, expression }) => {
    el.setAttribute('x-ignore', true)

    const asyncData = await import(expression)
    if (asyncData) {
        el.removeAttribute('x-ignore')
        Alpine.addScopeToNode(el, asyncData)
    }

}).before('ignore')
```

```
<div x-async-data="assets/component/comp-data-to-import.js">
...rest of the html...
</div>
```

**Reply by ekwoka** [reply]
That would definitely do it.

**Reply by Tim-Wils** [reply, marked as answer]
I just needed this for a project, so I went for it and made it work with this directive that also supports arguments:
```js
const asyncAlpineModulesPath = './components/[name].js'
const asyncAlpineModules = {}
Alpine.directive('async-data', async (el, { value, modifiers, expression }) => {
    const moduleName = expression.split('(')[0];
    if (!asyncAlpineModules[moduleName]) {
        const asyncData = await import(/* @vite-ignore */ asyncAlpineModulesPath.replace('[name]',moduleName))
        if (asyncData && asyncData.data) {
            asyncAlpineModules[moduleName] = true
            Alpine.data(moduleName, (...args) => {
                return asyncData.data(...args)
            })
        }
    }

    Alpine.mutateDom(()=> {
        el.setAttribute('x-data', expression)
        el.removeAttribute('x-async-data')
        el.removeAttribute('x-ignore')
        el._x_ignore = false
 
        Alpine.initTree(el)

    })
}).before('ignore')
```
```html
Elements:
<div x-data="{parentitem:true}">
<div x-ignore x-data x-async-data="test('first','second')">
<span x-text="item.test"></span>
</div>
</div>
```
My test.js is in components/test.js.

This is the cleanest I could get it for now. If anyone has some better adjustments, please let me know!

**Reply by Briareos17** [reply]
How do you splitting alpine components on build step to make it available for import from directive?
